safenode/src/protocol/types/transfers/mod.rs

```diff
@@ -0,0 +1,225 @@
+    pub recipients: Vec<(Token, DbcIdSource)>,
+    /// Any surplus amount after spending the necessary input dbcs.
+    pub change: (Token, PublicAddress),
+}
+
+/// The token transfer results.
+#[derive(Debug)]
+pub struct Outputs {
+    /// The dbcs that were created containing
+    /// the tokens sent to respective recipient.
+    pub created_dbcs: Vec<CreatedDbc>,
+    /// The dbc holding surplus tokens after
+    /// spending the necessary input dbcs.
+    pub change_dbc: Option<Dbc>,
+}
+
+/// A resulting dbc from a token transfer.
+#[derive(Debug, Clone)]
+pub struct CreatedDbc {
+    /// The dbc that was created.
+    pub dbc: Dbc,
+    /// This is useful for the sender to know how much they sent to each recipient.
+    /// They can't know this from the dbc itself, as the amount is encrypted.
+    pub amount: RevealedAmount,
+}
+
+/// A function for creating an offline transfer of tokens.
+/// This is done by creating new dbcs to the recipients (and a change dbc if any)
+/// by selecting from the available input dbcs, and creating the necessary
+/// spends to do so.
+///
+/// Those signed spends are found in each new dbc, and must be uploaded to the network
+/// for the transactions to take effect.
+/// The peers will validate each signed spend they receive, before accepting it.
+/// Once enough peers have accepted all the spends of the transaction, and serve
+/// them upon request, the transaction will be completed.
+///
+/// (Disabled for now: Transfer fees will be added if not in data-network.)
+/// (Disabled for now: DbcReason, can be added later.)
+#[allow(clippy::result_large_err)]
+pub fn create_transfer(
+    dbcs: Vec<(Dbc, DerivedKey)>,
+    recipients: Vec<(Token, DbcIdSource)>,
+    change_to: PublicAddress,
+) -> Result<Outputs> {
+    // We need to select the necessary number of dbcs from those that we were passed.
+    // This will also account for any fees.
+    let send_inputs = select_inputs(dbcs, recipients, change_to)?;
+    transfer(send_inputs)
+}
+
+/// The tokens of the input dbcs will be transfered to the
+/// new dbcs (and a change dbc if any), which are returned from this function.
+/// This does not register the transaction in the network.
+/// To do that, the `signed_spends` of each new dbc, has to be uploaded
+/// to the network. When those same signed spends can be retrieved from
+/// enough peers in the network, the transaction will be completed.
+#[allow(clippy::result_large_err)]
+fn transfer(send_inputs: Inputs) -> Result<Outputs> {
+    let Inputs {
+        dbcs_to_spend,
+        recipients,
+        change: (change, change_to),
+    } = send_inputs;
+
+    let mut inputs = vec![];
+    for (dbc, derived_key) in dbcs_to_spend {
+        let revealed_amount = match dbc.revealed_amount(&derived_key) {
+            Ok(amount) => amount,
+            Err(err) => {
+                warn!("Ignoring dbc, as it didn't have the correct derived key: {err}");
```

Comment:
Well, it's a private function and it is only called with inputs that are gotten from the `select_inputs` fn, which has already filtered out dbcs that we don't own (which won't get passed in there from the beginning, as we filter that at wallet level as well..).

Anyway, so that warn log line is practically unreachable.

We could move around some code to produce that `InputHistory` struct already in the `select_inputs` function, as we at the end of that function compare in and out amount, and return an error if we don't have enough to spend.